File: internal/agent/tools.ts
import path from "node:path";
import { readFile } from "node:fs/promises";
import minimatch from "minimatch";
import type { ToolDefinition as AikitToolDefinition, ToolCall as AikitToolCall } from "@volpestyle/ai-kit-node";
import type { MessagePart } from "../runstore/session_models";
import { applyUnifiedDiff } from "../util/patch";
import { defaultWalkOptions, walkFiles } from "../util/files";
import { runCommand } from "../util/exec";
import { buildRepoMap } from "./context";

export type ToolKind = "read" | "write" | "exec" | "network";

export interface ToolDefinition {
  name: string;
  description: string;
  parameters?: Record<string, unknown>;
  kind: ToolKind;
  requiresApproval?: boolean;
  allowWithoutApproval?: boolean;
}

export interface ToolCall {
  id: string;
  name: string;
  input: string;
}

export interface ToolResult {
  id: string;
  ok: boolean;
  parts: MessagePart[];
  artifacts?: string[];
  error?: string;
}

export interface Tool {
  definition(): ToolDefinition;
  invoke(call: ToolCall, signal?: AbortSignal): Promise<ToolResult>;
}

export interface ToolRegistry {
  definitions(): ToolDefinition[];
  invoke(call: ToolCall, signal?: AbortSignal): Promise<ToolResult>;
  get(name: string): Tool | undefined;
  add(tool: Tool): void;
}

export class Registry implements ToolRegistry {
  private tools = new Map<string, Tool>();

  constructor(...tools: Tool[]) {
    for (const tool of tools) {
      this.add(tool);
    }
  }

  add(tool: Tool): void {
    if (!tool) return;
    this.tools.set(tool.definition().name, tool);
  }

  definitions(): ToolDefinition[] {
    return Array.from(this.tools.values())
      .map((tool) => tool.definition())
      .sort((a, b) => a.name.localeCompare(b.name));
  }

  async invoke(call: ToolCall, signal?: AbortSignal): Promise<ToolResult> {
    const tool = this.tools.get(call.name);
    if (!tool) {
      return { id: call.id, ok: false, error: "unknown tool", parts: [] };
    }
    return tool.invoke(call, signal);
  }

  get(name: string): Tool | undefined {
    return this.tools.get(name);
  }
}

export function defaultToolRegistry(workspace: string, commands: string[]): ToolRegistry {
  const verifyCommands = commands.length ? commands : ["make test"];
  return new Registry(
    new RepoTreeTool(workspace, 500),
    new RepoMapTool(workspace, 400),
    new ReadFileTool(workspace, 400),
    new SearchTool(workspace, 50),
    new GitStatusTool(workspace),
    new ApplyPatchTool(workspace),
    new ShellTool(workspace, 30 * 60_000),
    new DiagramTool(workspace),
    new VerifyTool(workspace, verifyCommands, 30 * 60_000),
  );
}

export class AikitAdapter {
  toAikitTools(defs: ToolDefinition[]): AikitToolDefinition[] {
    return defs.map((def) => ({
      name: def.name,
      description: def.description,
      parameters:
        def.parameters ?? {
          type: "object",
          properties: {},
        },
    }));
  }

  fromAikitCall(call: AikitToolCall): ToolCall {
    const raw = call.argumentsJson?.trim() || "{}";
    return { id: call.id, name: call.name, input: raw };
  }
}

function safeWorkspacePath(workspace: string, rel: string): string {
  if (!rel.trim()) throw new Error("path is empty");
  const root = path.resolve(workspace);
  const abs = path.resolve(root, rel);
  const relPath = path.relative(root, abs);
  if (relPath === ".." || relPath.startsWith(`..${path.sep}`)) {
    throw new Error(`path escapes workspace: ${rel}`);
  }
  return abs;
}

function toJson(value: unknown): string {
  return JSON.stringify(value, null, 2);
}

export class RepoTreeTool implements Tool {
  constructor(private workspace: string, private maxFiles: number) {}

  definition(): ToolDefinition {
    return {
      name: "repo_tree",
      description: "List files in the workspace (relative paths).",
      kind: "read",
      parameters: {
        type: "object",
        properties: {
          max_files: { type: "integer" },
        },
      },
    };
  }

  async invoke(call: ToolCall): Promise<ToolResult> {
    let maxFiles = this.maxFiles;
    try {
      const parsed = JSON.parse(call.input || "{}") as { max_files?: number };
      if (parsed.max_files && parsed.max_files > 0) maxFiles = parsed.max_files;
    } catch {
      // ignore
    }
    const files = await walkFiles(this.workspace, defaultWalkOptions());
    const slice = maxFiles > 0 ? files.slice(0, maxFiles) : files;
    const text = slice.length ? slice.join("\n") : "workspace contains no files";
    return { id: call.id, ok: true, parts: [{ type: "text", text }] };
  }
}

export class RepoMapTool implements Tool {
  constructor(private workspace: string, private maxSymbols: number) {}

  definition(): ToolDefinition {
    return {
      name: "repo_map",
      description: "List symbols in the repo (Go/Python/JS/TS).",
      kind: "read",
      parameters: {
        type: "object",
        properties: {
          max_symbols: { type: "integer" },
        },
      },
    };
  }

  async invoke(call: ToolCall): Promise<ToolResult> {
    let maxSymbols = this.maxSymbols;
    try {
      const parsed = JSON.parse(call.input || "{}") as { max_symbols?: number };
      if (parsed.max_symbols && parsed.max_symbols > 0) maxSymbols = parsed.max_symbols;
    } catch {
      // ignore
    }
    if (maxSymbols <= 0) maxSymbols = 400;
    const files = await walkFiles(this.workspace, defaultWalkOptions());
    const out = buildRepoMap(this.workspace, files, maxSymbols);
    return { id: call.id, ok: true, parts: [{ type: "text", text: out }] };
  }
}

export class ReadFileTool implements Tool {
  constructor(private workspace: string, private maxLines: number) {}

  definition(): ToolDefinition {
    return {
      name: "read_file",
      description: "Read a file from the workspace with optional line range.",
      kind: "read",
      parameters: {
        type: "object",
        properties: {
          path: { type: "string" },
          start_line: { type: "integer" },
          end_line: { type: "integer" },
        },
        required: ["path"],
      },
    };
  }

  async invoke(call: ToolCall): Promise<ToolResult> {
    let input: { path: string; start_line?: number; end_line?: number };
    try {
      input = JSON.parse(call.input || "{}") as { path: string; start_line?: number; end_line?: number };
    } catch {
      return { id: call.id, ok: false, error: "invalid input", parts: [] };
    }
    const abs = safeWorkspacePath(this.workspace, input.path);
    const content = await readFile(abs, "utf8");
    const lines = content.split("\n");
    let start = input.start_line && input.start_line > 0 ? input.start_line : 1;
    let end = input.end_line && input.end_line > 0 ? input.end_line : lines.length;
    if (start < 1) start = 1;
    if (end > lines.length) end = lines.length;
    if (start > end) start = end;
    if (this.maxLines > 0 && end - start + 1 > this.maxLines) {
      end = Math.min(lines.length, start + this.maxLines - 1);
    }
    const snippet = lines.slice(start - 1, end).join("\n");
    return { id: call.id, ok: true, parts: [{ type: "text", text: snippet }] };
  }
}

export class SearchTool implements Tool {
  constructor(private workspace: string, private maxResults: number) {}

  definition(): ToolDefinition {
    return {
      name: "search",
      description: "Search for a substring in files.",
      kind: "read",
      parameters: {
        type: "object",
        properties: {
          query: { type: "string" },
          glob: { type: "string" },
          max_results: { type: "integer" },
        },
        required: ["query"],
      },
    };
  }

  async invoke(call: ToolCall): Promise<ToolResult> {
    let input: { query: string; glob?: string; max_results?: number };
    try {
      input = JSON.parse(call.input || "{}") as { query: string; glob?: string; max_results?: number };
    } catch {
      return { id: call.id, ok: false, error: "invalid input", parts: [] };
    }
    const query = input.query?.trim();
    if (!query) {
      return { id: call.id, ok: false, error: "query required", parts: [] };
    }
    let maxResults = this.maxResults;
    if (input.max_results && input.max_results > 0) maxResults = input.max_results;
    if (maxResults <= 0) maxResults = 50;
    const files = await walkFiles(this.workspace, defaultWalkOptions());
    const matches: string[] = [];
    for (const rel of files) {
      if (matches.length >= maxResults) break;
      if (input.glob && !minimatch(path.basename(rel), input.glob)) continue;
      const abs = path.join(this.workspace, rel);
      let content: string;
      try {
        content = await readFile(abs, "utf8");
      } catch {
        continue;
      }
      const lines = content.split("\n");
      for (let i = 0; i < lines.length; i++) {
        if (matches.length >= maxResults) break;
        if (lines[i].includes(query)) {
          matches.push(`${rel}:${i + 1}:${lines[i].trim()}`);
        }
      }
    }
    return { id: call.id, ok: true, parts: [{ type: "text", text: matches.join("\n") }] };
  }
}

export class GitStatusTool implements Tool {
  constructor(private workspace: string) {}

  definition(): ToolDefinition {
    return {
      name: "git_status",
      description: "Return git status --porcelain for the workspace.",
      kind: "read",
      parameters: { type: "object", properties: {} },
    };
  }

  async invoke(call: ToolCall): Promise<ToolResult> {
    try {
      const res = await runCommand("git status --porcelain", { dir: this.workspace, timeoutMs: 10_000 });
      return { id: call.id, ok: true, parts: [{ type: "text", text: res.stdout.trim() }] };
    } catch (err: unknown) {
      const result = (err as { result?: { stdout?: string; stderr?: string } }).result;
      return { id: call.id, ok: false, error: (err as Error).message, parts: [{ type: "text", text: result?.stdout ?? "" }] };
    }
  }
}

export class ApplyPatchTool implements Tool {
  constructor(private workspace: string) {}

  definition(): ToolDefinition {
    return {
      name: "apply_patch",
      description: "Apply a unified diff patch using git apply.",
      kind: "write",
      requiresApproval: true,
      parameters: {
        type: "object",
        properties: {
          patch: { type: "string" },
        },
        required: ["patch"],
      },
    };
  }

  async invoke(call: ToolCall): Promise<ToolResult> {
    let input: { patch: string };
    try {
      input = JSON.parse(call.input || "{}") as { patch: string };
    } catch {
      return { id: call.id, ok: false, error: "invalid input", parts: [] };
    }
    try {
      const res = await applyUnifiedDiff(this.workspace, input.patch);
      return { id: call.id, ok: true, parts: [{ type: "text", text: toJson(res) }] };
    } catch (err: unknown) {
      const result = (err as { result?: unknown }).result ?? { applied: false };
      return {
        id: call.id,
        ok: false,
        error: (err as Error).message,
        parts: [{ type: "text", text: toJson(result) }],
      };
    }
  }
}

export class ShellTool implements Tool {
  constructor(private workspace: string, private timeoutMs: number) {}

  definition(): ToolDefinition {
    return {
      name: "shell",
      description: "Run a shell command in the workspace.",
      kind: "exec",
      requiresApproval: true,
      parameters: {
        type: "object",
        properties: {
          command: { type: "string" },
          timeout_seconds: { type: "integer" },
        },
        required: ["command"],
      },
    };
  }

  async invoke(call: ToolCall): Promise<ToolResult> {
    let input: { command: string; timeout_seconds?: number };
    try {
      input = JSON.parse(call.input || "{}") as { command: string; timeout_seconds?: number };
    } catch {
      return { id: call.id, ok: false, error: "invalid input", parts: [] };
    }
    const timeout = input.timeout_seconds ? input.timeout_seconds * 1000 : this.timeoutMs;
    try {
      const res = await runCommand(input.command, { dir: this.workspace, timeoutMs: timeout });
      return { id: call.id, ok: true, parts: [{ type: "text", text: toJson(res) }] };
    } catch (err: unknown) {
      const result = (err as { result?: unknown }).result ?? {};
      return {
        id: call.id,
        ok: false,
        error: (err as Error).message,
        parts: [{ type: "text", text: toJson(result) }],
      };
    }
  }
}

export class DiagramTool implements Tool {
  constructor(private workspace: string) {}

  definition(): ToolDefinition {
    return {
      name: "diagram",
      description: "Render diagrams using make diagrams.",
      kind: "exec",
      requiresApproval: true,
      parameters: { type: "object", properties: {} },
    };
  }

  async invoke(call: ToolCall): Promise<ToolResult> {
    try {
      const res = await runCommand("make diagrams", { dir: this.workspace, timeoutMs: 30 * 60_000 });
      return { id: call.id, ok: true, parts: [{ type: "text", text: toJson(res) }] };
    } catch (err: unknown) {
      const result = (err as { result?: unknown }).result ?? {};
      return {
        id: call.id,
        ok: false,
        error: (err as Error).message,
        parts: [{ type: "text", text: toJson(result) }],
      };
    }
  }
}

export class VerifyTool implements Tool {
  constructor(private workspace: string, private commands: string[], private timeoutMs: number) {}

  definition(): ToolDefinition {
    return {
      name: "verify",
      description: "Run verification commands.",
      kind: "exec",
      parameters: { type: "object", properties: {} },
    };
  }

  async invoke(call: ToolCall): Promise<ToolResult> {
    if (!this.commands.length) this.commands = ["make test"];
    const results: Array<Record<string, unknown>> = [];
    let ok = true;
    for (const cmd of this.commands) {
      try {
        const res = await runCommand(cmd, { dir: this.workspace, timeoutMs: this.timeoutMs });
        results.push({
          cmd: res.cmd,
          exit_code: res.exit_code,
          stdout: res.stdout,
          stderr: res.stderr,
          duration: res.duration,
        });
      } catch (err: unknown) {
        ok = false;
        const result = (err as { result?: Record<string, unknown> }).result;
        results.push({
          cmd,
          exit_code: (result as { exit_code?: number })?.exit_code ?? 1,
          stdout: (result as { stdout?: string })?.stdout ?? "",
          stderr: (result as { stderr?: string })?.stderr ?? (err as Error).message,
          duration: (result as { duration?: string })?.duration ?? "",
        });
      }
    }
    const out = toJson(results);
    if (!ok) {
      return { id: call.id, ok: false, error: "verification failed", parts: [{ type: "text", text: out }] };
    }
    return { id: call.id, ok: true, parts: [{ type: "text", text: out }] };
  }
}

File: internal/agent/tools.py
from __future__ import annotations

import json
import os
import fnmatch
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional, Protocol

from internal.runstore.session_models import MessagePart
from internal.util.exec import ExecOptions, run_command
from internal.util.files import default_walk_options, walk_files
from internal.util.patch import apply_unified_diff
from .context import build_repo_map

ToolKind = str


@dataclass
class ToolDefinition:
    name: str
    description: str
    kind: ToolKind
    parameters: Optional[Dict[str, object]] = None
    requires_approval: bool = False
    allow_without_approval: bool = False


@dataclass
class ToolCall:
    id: str
    name: str
    input: str


@dataclass
class ToolResult:
    id: str
    ok: bool
    parts: List[MessagePart]
    artifacts: Optional[List[str]] = None
    error: Optional[str] = None


class Tool(Protocol):
    def definition(self) -> ToolDefinition:
        ...

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        ...


class ToolRegistry(Protocol):
    def definitions(self) -> List[ToolDefinition]:
        ...

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        ...

    def get(self, name: str) -> Optional[Tool]:
        ...

    def add(self, tool: Tool) -> None:
        ...


class Registry:
    def __init__(self, *tools: Tool) -> None:
        self._tools: Dict[str, Tool] = {}
        for tool in tools:
            self.add(tool)

    def add(self, tool: Tool) -> None:
        if not tool:
            return
        self._tools[tool.definition().name] = tool

    def definitions(self) -> List[ToolDefinition]:
        return sorted(
            (tool.definition() for tool in self._tools.values()),
            key=lambda d: d.name,
        )

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        tool = self._tools.get(call.name)
        if not tool:
            return ToolResult(id=call.id, ok=False, error="unknown tool", parts=[])
        return tool.invoke(call, signal)

    def get(self, name: str) -> Optional[Tool]:
        return self._tools.get(name)


def default_tool_registry(workspace: str, commands: List[str]) -> ToolRegistry:
    verify_commands = commands or ["make test"]
    return Registry(
        RepoTreeTool(workspace, 500),
        RepoMapTool(workspace, 400),
        ReadFileTool(workspace, 400),
        SearchTool(workspace, 50),
        GitStatusTool(workspace),
        ApplyPatchTool(workspace),
        ShellTool(workspace, 30 * 60_000),
        DiagramTool(workspace),
        VerifyTool(workspace, verify_commands, 30 * 60_000),
    )


class AikitAdapter:
    def to_aikit_tools(self, defs: List[ToolDefinition]):
        from ai_kit import ToolDefinition as AikitToolDefinition

        out = []
        for definition in defs:
            out.append(
                AikitToolDefinition(
                    name=definition.name,
                    description=definition.description,
                    parameters=definition.parameters
                    or {
                        "type": "object",
                        "properties": {},
                    },
                )
            )
        return out

    def from_aikit_call(self, call) -> ToolCall:
        raw = _normalize_tool_input(getattr(call, "argumentsJson", "") or "")
        return ToolCall(id=call.id, name=call.name, input=raw)


def _safe_workspace_path(workspace: str, rel: str) -> str:
    if not rel.strip():
        raise ValueError("path is empty")
    root = Path(workspace).resolve()
    abs_path = (root / rel).resolve()
    rel_path = os.path.relpath(abs_path, root)
    if rel_path == ".." or rel_path.startswith(".." + os.sep):
        raise ValueError(f"path escapes workspace: {rel}")
    return str(abs_path)


def _to_json(value: object) -> str:
    return json.dumps(value, indent=2, default=str)


def _normalize_tool_input(raw: str) -> str:
    trimmed = raw.strip()
    if not trimmed:
        return "{}"
    if _is_valid_json(trimmed):
        return trimmed
    candidate = _extract_last_json_object(trimmed)
    if candidate and _is_valid_json(candidate):
        return candidate
    return trimmed


def _is_valid_json(value: str) -> bool:
    try:
        json.loads(value)
        return True
    except Exception:
        return False


def _extract_last_json_object(value: str) -> Optional[str]:
    depth = 0
    end = -1
    for i in range(len(value) - 1, -1, -1):
        ch = value[i]
        if ch == "}":
            if depth == 0:
                end = i
            depth += 1
            continue
        if ch == "{":
            if depth > 0:
                depth -= 1
                if depth == 0 and end != -1:
                    return value[i : end + 1]
    return None


class RepoTreeTool:
    def __init__(self, workspace: str, max_files: int) -> None:
        self._workspace = workspace
        self._max_files = max_files

    def definition(self) -> ToolDefinition:
        return ToolDefinition(
            name="repo_tree",
            description="List files in the workspace (relative paths).",
            kind="read",
            parameters={"type": "object", "properties": {"max_files": {"type": "integer"}}},
        )

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        max_files = self._max_files
        try:
            payload = json.loads(call.input or "{}")
            if payload.get("max_files"):
                max_files = int(payload.get("max_files"))
        except Exception:
            pass
        files = walk_files(self._workspace, default_walk_options())
        slice_ = files[: max_files if max_files > 0 else None]
        text = "\n".join(slice_) if slice_ else "workspace contains no files"
        return ToolResult(id=call.id, ok=True, parts=[MessagePart(type="text", text=text)])


class RepoMapTool:
    def __init__(self, workspace: str, max_symbols: int) -> None:
        self._workspace = workspace
        self._max_symbols = max_symbols

    def definition(self) -> ToolDefinition:
        return ToolDefinition(
            name="repo_map",
            description="List symbols in the repo (Go/Python/JS/TS).",
            kind="read",
            parameters={"type": "object", "properties": {"max_symbols": {"type": "integer"}}},
        )

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        max_symbols = self._max_symbols
        try:
            payload = json.loads(call.input or "{}")
            if payload.get("max_symbols"):
                max_symbols = int(payload.get("max_symbols"))
        except Exception:
            pass
        if max_symbols <= 0:
            max_symbols = 400
        files = walk_files(self._workspace, default_walk_options())
        out = build_repo_map(self._workspace, files, max_symbols)
        return ToolResult(id=call.id, ok=True, parts=[MessagePart(type="text", text=out)])


class ReadFileTool:
    def __init__(self, workspace: str, max_lines: int) -> None:
        self._workspace = workspace
        self._max_lines = max_lines

    def definition(self) -> ToolDefinition:
        return ToolDefinition(
            name="read_file",
            description="Read a file from the workspace with optional line range.",
            kind="read",
            parameters={
                "type": "object",
                "properties": {
                    "path": {"type": "string"},
                    "start_line": {"type": "integer"},
                    "end_line": {"type": "integer"},
                },
                "required": ["path"],
            },
        )

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        try:
            payload = json.loads(call.input or "{}")
        except Exception:
            return ToolResult(id=call.id, ok=False, error="invalid input", parts=[])
        abs_path = _safe_workspace_path(self._workspace, payload.get("path", ""))
        content = Path(abs_path).read_text(encoding="utf-8", errors="ignore")
        lines = content.split("\n")
        start = int(payload.get("start_line") or 1)
        end = int(payload.get("end_line") or len(lines))
        if start < 1:
            start = 1
        if end > len(lines):
            end = len(lines)
        if start > end:
            start = end
        if self._max_lines > 0 and (end - start + 1) > self._max_lines:
            end = min(len(lines), start + self._max_lines - 1)
        snippet = "\n".join(lines[start - 1 : end])
        return ToolResult(id=call.id, ok=True, parts=[MessagePart(type="text", text=snippet)])


class SearchTool:
    def __init__(self, workspace: str, max_results: int) -> None:
        self._workspace = workspace
        self._max_results = max_results

    def definition(self) -> ToolDefinition:
        return ToolDefinition(
            name="search",
            description="Search for a substring in files.",
            kind="read",
            parameters={
                "type": "object",
                "properties": {
                    "query": {"type": "string"},
                    "glob": {"type": "string"},
                    "max_results": {"type": "integer"},
                },
                "required": ["query"],
            },
        )

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        try:
            payload = json.loads(call.input or "{}")
        except Exception:
            return ToolResult(id=call.id, ok=False, error="invalid input", parts=[])
        query = (payload.get("query") or "").strip()
        if not query:
            return ToolResult(id=call.id, ok=False, error="query required", parts=[])
        max_results = int(payload.get("max_results") or self._max_results)
        if max_results <= 0:
            max_results = 50
        glob = payload.get("glob")
        files = walk_files(self._workspace, default_walk_options())
        matches: List[str] = []
        for rel in files:
            if len(matches) >= max_results:
                break
            if glob and not fnmatch.fnmatch(Path(rel).name, glob):
                continue
            abs_path = Path(self._workspace) / rel
            try:
                content = abs_path.read_text(encoding="utf-8", errors="ignore")
            except Exception:
                continue
            lines = content.split("\n")
            for idx, line in enumerate(lines, start=1):
                if len(matches) >= max_results:
                    break
                if query in line:
                    matches.append(f"{rel}:{idx}:{line.strip()}")
        return ToolResult(id=call.id, ok=True, parts=[MessagePart(type="text", text="\n".join(matches))])


class GitStatusTool:
    def __init__(self, workspace: str) -> None:
        self._workspace = workspace

    def definition(self) -> ToolDefinition:
        return ToolDefinition(
            name="git_status",
            description="Return git status --porcelain for the workspace.",
            kind="read",
            parameters={"type": "object", "properties": {}},
        )

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        try:
            res = run_command("git status --porcelain", ExecOptions(dir=self._workspace, timeout_ms=10_000, signal=signal))
            return ToolResult(id=call.id, ok=True, parts=[MessagePart(type="text", text=res.stdout.strip())])
        except Exception as err:
            result = getattr(err, "result", None)
            stdout = getattr(result, "stdout", "") if result else ""
            return ToolResult(id=call.id, ok=False, error=str(err), parts=[MessagePart(type="text", text=stdout)])


class ApplyPatchTool:
    def __init__(self, workspace: str) -> None:
        self._workspace = workspace

    def definition(self) -> ToolDefinition:
        return ToolDefinition(
            name="apply_patch",
            description="Apply a unified diff patch using git apply.",
            kind="write",
            requires_approval=True,
            parameters={
                "type": "object",
                "properties": {"patch": {"type": "string"}},
                "required": ["patch"],
            },
        )

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        try:
            payload = json.loads(call.input or "{}")
        except Exception:
            return ToolResult(id=call.id, ok=False, error="invalid input", parts=[])
        try:
            result = apply_unified_diff(self._workspace, payload.get("patch", ""), signal=signal)
            return ToolResult(id=call.id, ok=True, parts=[MessagePart(type="text", text=_to_json(result.__dict__))])
        except Exception as err:
            result = getattr(err, "result", None)
            data = result.__dict__ if result else {"applied": False}
            return ToolResult(id=call.id, ok=False, error=str(err), parts=[MessagePart(type="text", text=_to_json(data))])


class ShellTool:
    def __init__(self, workspace: str, timeout_ms: int) -> None:
        self._workspace = workspace
        self._timeout_ms = timeout_ms

    def definition(self) -> ToolDefinition:
        return ToolDefinition(
            name="shell",
            description="Run a shell command in the workspace.",
            kind="exec",
            requires_approval=True,
            parameters={
                "type": "object",
                "properties": {
                    "command": {"type": "string"},
                    "timeout_seconds": {"type": "integer"},
                },
                "required": ["command"],
            },
        )

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        try:
            payload = json.loads(call.input or "{}")
        except Exception:
            return ToolResult(id=call.id, ok=False, error="invalid input", parts=[])
        timeout = int(payload.get("timeout_seconds") or 0)
        timeout_ms = timeout * 1000 if timeout else self._timeout_ms
        try:
            res = run_command(payload.get("command", ""), ExecOptions(dir=self._workspace, timeout_ms=timeout_ms, signal=signal))
            return ToolResult(id=call.id, ok=True, parts=[MessagePart(type="text", text=_to_json(res.__dict__))])
        except Exception as err:
            result = getattr(err, "result", None)
            data = result.__dict__ if result else {}
            return ToolResult(id=call.id, ok=False, error=str(err), parts=[MessagePart(type="text", text=_to_json(data))])


class DiagramTool:
    def __init__(self, workspace: str) -> None:
        self._workspace = workspace

    def definition(self) -> ToolDefinition:
        return ToolDefinition(
            name="diagram",
            description="Render diagrams using make diagrams.",
            kind="exec",
            requires_approval=True,
            parameters={"type": "object", "properties": {}},
        )

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        try:
            res = run_command("make diagrams", ExecOptions(dir=self._workspace, timeout_ms=30 * 60_000, signal=signal))
            return ToolResult(id=call.id, ok=True, parts=[MessagePart(type="text", text=_to_json(res.__dict__))])
        except Exception as err:
            result = getattr(err, "result", None)
            data = result.__dict__ if result else {}
            return ToolResult(id=call.id, ok=False, error=str(err), parts=[MessagePart(type="text", text=_to_json(data))])


class VerifyTool:
    def __init__(self, workspace: str, commands: List[str], timeout_ms: int) -> None:
        self._workspace = workspace
        self._commands = commands
        self._timeout_ms = timeout_ms

    def definition(self) -> ToolDefinition:
        return ToolDefinition(
            name="verify",
            description="Run verification commands.",
            kind="exec",
            parameters={"type": "object", "properties": {}},
        )

    def invoke(self, call: ToolCall, signal=None) -> ToolResult:
        if not self._commands:
            self._commands = ["make test"]
        results: List[Dict[str, object]] = []
        ok = True
        for cmd in self._commands:
            try:
                res = run_command(cmd, ExecOptions(dir=self._workspace, timeout_ms=self._timeout_ms, signal=signal))
                results.append(
                    {
                        "cmd": res.cmd,
                        "exit_code": res.exit_code,
                        "stdout": res.stdout,
                        "stderr": res.stderr,
                        "duration": res.duration,
                    }
                )
            except Exception as err:
                ok = False
                result = getattr(err, "result", None)
                results.append(
                    {
                        "cmd": cmd,
                        "exit_code": getattr(result, "exit_code", 1) if result else 1,
                        "stdout": getattr(result, "stdout", "") if result else "",
                        "stderr": getattr(result, "stderr", "") if result else str(err),
                        "duration": getattr(result, "duration", "") if result else "",
                    }
                )
        out = _to_json(results)
        if not ok:
            return ToolResult(id=call.id, ok=False, error="verification failed", parts=[MessagePart(type="text", text=out)])
        return ToolResult(id=call.id, ok=True, parts=[MessagePart(type="text", text=out)])
